### ISSUE TITLE
[Windows] Having weird issues on next boots after closing content with Flycast.

### ISSUE THREAD

**Issue by bslenul** [libretro/RetroArch#11940]
Initially posted on Flycast repo (https://github.com/libretro/flycast/issues/1019) but to quote barbudreadmon:

> Meaning it's probably somehow related to a compiler bug :/

So I'm reposting here!

## Description

Hey!

Short version:

* 2nd boot is sped up if I have refresh rate set to 144Hz in RA.
* 3rd boot always give me slow and distorted audio, no matter if refresh rate is set to 60 or 144Hz.
* When loading a .m3u the discs are "re-added" on next boots, so loading a 2 discs game it will have 4 discs on 2nd boot, 6 discs on 3rd boot, 8 discs on 4th boot, etc.
* Seems to be Windows only, with both gl/core and vulkan.
* Doesn't happen if I build RA myself (but vulkan segfault on close content in this case).
* Other HW rendered cores work fine (Beetle PSX HW, Dolphin, PPSSPP, etc.).

Long version:

So I'm getting super weird issues after closing content, basically it goes like this:

* 1st launch: no issue, video and audio are fine.

https://user-images.githubusercontent.com/33353403/104902584-32a00a80-597f-11eb-923a-48751474e5a9.mp4

* Close content and launch again: video is sped up, audio is fine.

https://user-images.githubusercontent.com/33353403/104902606-37fd5500-597f-11eb-916d-d19b65e664df.mp4

* Close content and launch again: this time video is fine but audio is too slow and distorted.

https://user-images.githubusercontent.com/33353403/104902618-3d5a9f80-597f-11eb-9389-bbddce123fd9.mp4

Also I was playing a 2 discs game using .m3u and again no issue on first launch, the core sees 2 discs, then close content and launch again, the core now thinks it's a 4 discs game:

![image](https://user-images.githubusercontent.com/33353403/105701717-03ebdc00-5f0b-11eb-8da6-408a3d65b5f8.png)

And yeah in the "Disc Control" I have both discs x2 (then x3 on next boot, etc.):

![image](https://user-images.githubusercontent.com/33353403/105701761-19610600-5f0b-11eb-8f64-5f3a2bd337b7.png)

Up-to-date core (https://github.com/libretro/flycast/commit/ea1278ee1faf31068ff3b9122779ad5d771d88cc) and latest RA nightly (b9691597d1c2996008417e905dac74a9d7b13629), I haven't found anything special in it but here's the log: https://pastebin.com/TUMuvKmx

I'm not entirely sure if it's a core issue or a RetroArch issue because for some reasons it does NOT happen if I build RA myself (with gl/core at least, vulkan gives me a segfault on close content when I build RA myself). But in the other hand it seems to be the only HW rendered core where this happens :/ 

It happens on Windows with gl/core and vulkan, I wasn't able to reproduce on my Linux VM (with gl/core at least, can't use vulkan on my VM). The times where the video is sped up seems to only happen when I have 144Hz set in RA, the slow audio happens with both 60Hz and 144Hz.

In downloaded RA, you can see in the videos above that framerate seems to be locked at 120 on 2nd attempt, on the 3rd attempt it's around ~130. Haven't recorded but with my own build of RA, framerate counter shows a locked 60 on every attempt.

For the comparison between downloaded and built versions of RA, I made sure to use the same retroarch.cfg and the same Flycast.opt files! So it shouldn't be a different setting/core option causing this. Same commit for RA and same version of the core of course.

### Expected behavior

Playing at normal speed on next boots after closing content and .m3u not being added to the previous launch.

### Actual behavior

See description.

### Steps to reproduce the bug

1. Boot a DC game.
2. Close content.
3. Boot the game again.

### Bisect Results

That's my big problem, as I said it doesn't happen when I build RA myself, and on the buildbot the latest version I can get is from 1-12 which is not old enough. Worked fine on stable release so that's a big period: it happened between 8-11 and 1-12, yay! :D

### Version/Commit

- RetroArch: 1.9.0 / b9691597d1c2996008417e905dac74a9d7b13629

### Environment information

- OS: Windows 10, like I said, I wasn't able to reproduce on my Linux VM. It's also not happening on my 32bit Android phone.

**Comment by bslenul** [comment]
Ah, noticed something else strange:

https://user-images.githubusercontent.com/33353403/106211318-e350aa00-61c8-11eb-9016-ab488835ec96.mp4

I'm loading MvC2 first, as usual no issue on first boot and you can see that the sprites are normal, then I close content and I load Ikaruga... except it's loading MvC2 again and the sprites are completely messed up (+ running too fast as explained previously) 😅 

I didn't show it in the video but on the 2nd boot even if it loads MvC2 instead of Ikaruga, the VMU file `Ikaruga (Japan).A1.bin` is being created in my saves folder... with a MvC2 save inside.

**Comment by bslenul** [comment]
Oh, while doing some tests I made an interesting discovery: I noticed that with glcore on a fresh RA install it doesn't happen, I can boot the core how many times I want the speed issue doesn't happen.

But then, if I switch to vulkan it crashes on close content, then I launched it again multiple times and after a few tries it doesn't crash anymore (magic? 🤷 ), and from that point the bug is triggered, now no matter if I'm using glcore or vulkan I'll get the bugs I described previously on successive boots...

And even more strange, if I rename the RA folder then it's back to square one, the bug won't happen again and vulkan will crash on close content, until I trigger the bug again...

This makes absolutely NO SENSE to me, and it reminds me A LOT of that melonDS issue: https://github.com/libretro/melonDS/issues/98.

**Comment by bslenul** [comment]
Kinda make sense with all the issues I described I guess, but basically it looks like the core doesn't fully "unload" or something after closing content, like for example I can't delete the .dll from my cores folder after closing content, it says it's being used until I close RetroArch.

**Comment by bslenul** [comment]
So yeah, just found a program that can unload a dll from its process:

https://user-images.githubusercontent.com/33353403/107884109-35950900-6ef3-11eb-9048-5165a2306baf.mp4

Not unloading the dll -> speed issue + core being confused about what it's loading on consecutive launches.
Unloading the dll between 2 attempts -> works like a charm! No speed issue and the core loads the proper game and proper number of discs.

Still not 100% sure if it's a RA or a core issue (or Windows?), it's super strange and confusing...

**Comment by bslenul** [comment]
I talked about it on Discord but quick recap here:

* For some reason after 2 crashes on close content, Windows decides to create this registry key/value:
```
[HKEY_CURRENT_USER\SOFTWARE\Microsoft\Windows NT\CurrentVersion\AppCompatFlags\Layers]
"D:\\programmes\\RetroArch\\retroarch.exe"="$ IgnoreFreeLibrary<flycast_libretro.dll>"
```
* It seems to prevent Flycast from unloading at all on close content, which then causes the weird issues mentioned above.
* The registry key is created with RetroArch built using MXE, it's never created when using RA built with MinGW.
* Deleting the registry key will make RA crash on close content again, and it will be recreated on 2nd crash again.

**Comment by bslenul** [comment]
Closing this, fixed by updating to latest Nvidia drivers (465.89 when typing this), idk if it matters but I checked the "make a clean install" box during the installation. But yeah, no crash on close content, no registry key anymore and no weird behavior on successive loads :)
